File: public/code/smc__eq_sweep.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def eqSweep(c):
    if len(c) < 50:
        return make_signal(reason="Insufficient data")
    a = atr(c, 14)
    tol = a[len(a) - 1] * 0.15
    rh = [v for v in swingHighs(c, 5) if v != None][-10:]
    rl = [v for v in swingLows(c, 5) if v != None][-10:]
    eqh = None
    eql = None
    for i in range(0, len(rh) - 1):
        for j in range(i + 1, len(rh)):
            if abs(rh[i] - rh[j]) <= tol:
                eqh = max(rh[i], rh[j])
    for i in range(0, len(rl) - 1):
        for j in range(i + 1, len(rl)):
            if abs(rl[i] - rl[j]) <= tol:
                eql = min(rl[i], rl[j])
    cur = c[len(c) - 1]
    prev = c[len(c) - 2]
    if eql != None  and  prev['low'] < eql  and  prev['close'] > eql  and  cur['close'] > cur['open']:
        slv = prev['low'] - 0.3 * a[len(a) - 1]
        r = cur['close'] - slv
        return make_signal(signal="long", entry=cur['close'], stop_loss=slv, take_profit=[cur['close'] + r * 2, cur['close'] + r * 3, cur['close'] + r * 5], confidence=0.76, reason=f"EQL sweep @ {eql}")
    if eqh != None  and  prev['high'] > eqh  and  prev['close'] < eqh  and  cur['close'] < cur['open']:
        slv = prev['high'] + 0.3 * a[len(a) - 1]
        r = slv - cur['close']
        return make_signal(signal="short", entry=cur['close'], stop_loss=slv, take_profit=[cur['close'] - r * 2, cur['close'] - r * 3, cur['close'] - r * 5], confidence=0.76, reason=f"EQH sweep @ {eqh}")
    return make_signal(reason="EQH/EQL sweep yok")

File: public/code/smc__eq_sweep.ts
import { Candle, Signal, makeSignal, ema, atr, swingHighs, swingLows } from "../indicators";

export function eqSweep(c: Candle[]): Signal {
  if (c.length < 50) return makeSignal({ reason: "Insufficient data" });
  const a = atr(c, 14), tol = a[a.length - 1] * 0.15;
  const rh = swingHighs(c, 5).filter((v): v is number => v !== null).slice(-10);
  const rl = swingLows(c, 5).filter((v): v is number => v !== null).slice(-10);
  let eqh: number | null = null, eql: number | null = null;
  for (let i = 0; i < rh.length - 1; i++) for (let j = i + 1; j < rh.length; j++) if (Math.abs(rh[i] - rh[j]) <= tol) eqh = Math.max(rh[i], rh[j]);
  for (let i = 0; i < rl.length - 1; i++) for (let j = i + 1; j < rl.length; j++) if (Math.abs(rl[i] - rl[j]) <= tol) eql = Math.min(rl[i], rl[j]);
  const cur = c[c.length - 1], prev = c[c.length - 2];
  if (eql !== null && prev.low < eql && prev.close > eql && cur.close > cur.open) {
    const slv = prev.low - 0.3 * a[a.length - 1], r = cur.close - slv;
    return makeSignal({ signal: "long", entry: cur.close, stop_loss: slv, take_profit: [cur.close + r * 2, cur.close + r * 3, cur.close + r * 5], confidence: 0.76, reason: `EQL sweep @ ${eql.toFixed(2)}` });
  }
  if (eqh !== null && prev.high > eqh && prev.close < eqh && cur.close < cur.open) {
    const slv = prev.high + 0.3 * a[a.length - 1], r = slv - cur.close;
    return makeSignal({ signal: "short", entry: cur.close, stop_loss: slv, take_profit: [cur.close - r * 2, cur.close - r * 3, cur.close - r * 5], confidence: 0.76, reason: `EQH sweep @ ${eqh.toFixed(2)}` });
  }
  return makeSignal({ reason: "EQH/EQL sweep yok" });
}
